src/main/scala/mesosphere/marathon/api/forwarder/AsyncUrlConnectionRequestForwarder.scala

```diff
@@ -0,0 +1,210 @@
+      if !name.equalsIgnoreCase("content-length")
+      if !name.equalsIgnoreCase("user-agent")
+
+      headerValues <- Option(request.getHeaders(name))
+      headerValue <- headerValues.seq
+    } {
+      logger.debug(s"add RawHeader $name: $headerValue")
+      headers += RawHeader(name, headerValue)
+    }
+
+    headers += RawHeader(HEADER_VIA, viaValue)
+    val forwardedFor = Seq(
+      Option(request.getHeader(HEADER_FORWARDED_FOR)),
+      Option(request.getRemoteAddr)
+    ).flatten.mkString(",")
+    headers += RawHeader(HEADER_FORWARDED_FOR, forwardedFor)
+    Option(request.getHeader("User-Agent")).foreach { ua =>
+      headers += akka.http.scaladsl.model.headers.`User-Agent`(ua)
+    }
+    headers.result()
+  }
+
+  private def createAndConfigureConnection(asyncContext: AsyncContext, url: URL, clientRequest: HttpServletRequest): Future[HttpResponse] = {
+    val method = HttpMethods.getForKey(clientRequest.getMethod).getOrElse(HttpMethod.custom(clientRequest.getMethod))
+    val uri = Uri(url.toString)
+    val proxyHeaders = proxiedRequestHeaders(clientRequest)
+
+    val contentType = Option(clientRequest.getContentType).map(ContentType.parse) match {
+      case Some(Left(ex)) =>
+        logger.error(s"Ignoring unparseable Content-Type: ${clientRequest.getContentType}", ex)
+        ContentTypes.NoContentType
+      case Some(Right(contentType)) =>
+        contentType
+      case None =>
+        ContentTypes.NoContentType
+    }
+
+    val entity = clientRequest.getContentLengthLong match {
+      case 0 =>
+        HttpEntity.empty(contentType)
+      case -1 =>
+        HttpEntity.apply(contentType, Source.empty)
+      case contentLength =>
+        HttpEntity.apply(contentType, contentLength, ServletInputStreamSource.forAsyncContext(asyncContext))
+    }
+
+    val proxyRequest = HttpRequest(method, uri = uri, headers = proxyHeaders, entity = entity)
+    Http().singleRequest(request = proxyRequest, connectionContext = connectionContext, settings = poolSettings)
+  }
+
+  override def forward(url: URL, request: HttpServletRequest, response: HttpServletResponse): Unit = {
+    require(request.isAsyncSupported(), "ServletRequest does not support async mode")
+
+    logger.info(s"Proxying request to ${request.getMethod} $url from $myHostPort")
+
+    val asyncContext = request.startAsync()
+    asyncContext.setTimeout(0L) // delegate timeout to stream
+
+    val result: Future[Done] = try {
+      val hasProxyLoop: Boolean = Option(request.getHeaders(HEADER_VIA)).exists(_.seq.contains(viaValue))
+
+      if (hasProxyLoop) {
+        logger.error("Prevent proxy cycle, rejecting request")
+        response.sendError(BadGateway.intValue, ERROR_STATUS_LOOP)
+        Future.successful(Done)
+      } else {
+        val leaderRequest = try {
+          createAndConfigureConnection(asyncContext, url, request)
+        } catch {
+          case ex: Throwable =>
+            Future.failed(ex)
```

Comment:
Hm. I actually like to abort early. So instead of setting `result` to `Future.failed` I'd just do

```scala
logger.error("Unhandled proxy exception", e)
response.sendError(InternalServerError.intValue)
return
```

There is no need for a big try-catch block IMHO. What do you think?

Reply:
We need to close the async context. Returning a failed future funnels all of the logic through a single place.

Reply:
On review, this exception catching was unnecessary since we have a general catch-all later.